core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -404,4 +411,64 @@ private void updateBlockInStream(long blockId) throws IOException {
       mShouldCacheCurrentBlock = mAlluxioStorageType.isStore();
     }
   }
+
+  /**
+   * Seeks to a file position. Blocks are not cached unless they are fully read. This is only called
+   * by {@link FileInStream#seek}.
+   *
+   * @param pos The position to seek to. It is guaranteed to be valid (pos >= 0 && pos != mPos &&
+   *            pos <= mFileLength)
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekWithoutCachingIncompleteBlocks(long pos) throws IOException {
+    seekBlockInStream(pos);
+    checkAndAdvanceBlockInStream();
+    mCurrentBlockInStream.seek(mPos % mBlockSize);
+  }
+
+  /**
+   * Seeks to a file position. Blocks are cached even if they are fully read. This is only called by
+   * {@link FileInStream#seek}.
+   *
+   * @param pos The position to seek to. It is guaranteed to be valid (pos >= 0 && pos != mPos &&
+   *            pos <= mFileLength).
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekWithCachingIncompleteBlocks(long pos) throws IOException {
+    boolean isInCurrentBlock = pos / mBlockSize == mPos / mBlockSize;
+    if (mShouldCacheCurrentBlock) {
```

Comment:
I agree with you. I actually started with the way you are suggesting, but the code will be much longer and have some code duplication. The current logic is this:
1.  Cache the current block or read till pos depending on whether pos > mPos or pos is out of the current block.
2.  If pos is in current block, set mPos to pos. Otherwise, jump to pos and read the prefix of the new block. 

Basically, it is trying to maintain an invariant that we discussed in person.